// @babylonjs/core/Cameras/Limits/geospatialLimits.js
import { Epsilon } from "../../Maths/math.constants.js";
/**
 * Limits for geospatial camera
 */
export class GeospatialLimits {
    /**
     * @param planetRadius The radius of the planet
     */
    constructor(planetRadius) {
        this._radiusMin = Epsilon;
        this._radiusMax = Infinity;
        /** Gets the minimum pitch angle (angle from horizon) -- 0 means looking straight down at planet */
        this.pitchMin = Epsilon;
        /**  Gets the maximum pitch angle (angle from horizon) -- Pi/1 means looking at horizon */
        this.pitchMax = Math.PI / 2;
        /** Gets the minimum yaw angle (rotation around up axis) */
        this.yawMin = -Infinity;
        /** Gets the maximum yaw angle (rotation around up axis) */
        this.yawMax = Infinity;
        /**
         * Defines the distance used to consider the camera in pan mode vs pinch/zoom.
         * Basically if your fingers moves away from more than this distance you will be considered
         * in pinch mode.
         */
        this.pinchToPanMax = 20;
        this._planetRadius = planetRadius;
    }
    get radiusMin() {
        return this._radiusMin;
    }
    /**
     * Sets the minimum radius
     */
    set radiusMin(value) {
        this._radiusMin = value;
    }
    get radiusMax() {
        return this._radiusMax;
    }
    /**
     * Sets the maximum radius
     */
    set radiusMax(value) {
        this._radiusMax = value;
    }
    /**
     * Gets the planet radius used for altitude/radius conversions
     */
    get planetRadius() {
        return this._planetRadius;
    }
    /** Sets the planet radius and updates the radius limits to maintain current altitude */
    set planetRadius(value) {
        this._planetRadius = value;
    }
}
